transports/tls-quic/src/certificate.rs

```diff
@@ -0,0 +1,231 @@
+        // in the libp2p Public Key Extension.
+        // SignedKey ::= SEQUENCE {
+        //    publicKey BIT STRING,
+        //    signature BIT STRING
+        // }
+        let extension_content = {
+            let serialized_pubkey = keypair.public().into_protobuf_encoding();
+            yasna::encode_der(&(serialized_pubkey, signature))
+        };
+
+        // The libp2p Public Key Extension is a X.509 extension
+        // with the Object Identier 1.3.6.1.4.1.53594.1.1,
+        // allocated by IANA to the libp2p project at Protocol Labs.
+        let libp2p_ext_oid = &[1, 3, 6, 1, 4, 1, 53594, 1, 1];  // Based on libp2p TLS 1.3 specs
+
+        // This extension MAY be marked critical.
+        rcgen::CustomExtension::from_oid_content(libp2p_ext_oid, extension_content)
+    };
+
+    let certificate = {
+        let mut params = rcgen::CertificateParams::new(vec![]);
+        params.distinguished_name = rcgen::DistinguishedName::new();
+        params.custom_extensions.push(libp2p_extension);
+        params.alg = signature_algorithm;
+        params.key_pair = Some(certif_keypair);
+        rcgen::Certificate::from_params(params)?
+    };
+
+    Ok(certificate)
+}
+
+// The contents of the specific libp2p extension, containing the public host key
+// and a signature performed using the private host key
+pub struct P2pExtension {
+    public_key: identity::PublicKey,
+    signature: Vec<u8>,
+}
+
+pub struct P2pCertificate<'a> {
+    certificate: X509Certificate<'a>,
+    extension: P2pExtension,
+}
+
+/// Parse TLS certificate from DER input that includes a libp2p-specific
+/// certificate extension containing the public key of the given keypair.
+pub fn parse_certificate(der_input: &[u8]) -> Result<P2pCertificate, X509Error> {
+    let x509 = X509Certificate::from_der(der_input)
+        .map(|(_rest_input, x509)| x509)
+        .map_err(|e| {
+            use x509_parser::nom::Err::*;
+            match e {
+                Error(e) => e,
+                Failure(e) => e,
+                Incomplete(_) => X509Error::Generic, // should never happen
+            }
+        })?;
+
+    let p2p_ext_oid = der_parser::oid::Oid::from(&[1, 3, 6, 1, 4, 1, 53594, 1, 1])
+        .expect("This is a valid OID of p2p extension; qed");
+
+    let mut libp2p_extension = None;
+
+    for (oid, ext) in x509.extensions() {
+        if oid == &p2p_ext_oid && libp2p_extension.is_some() {
+            // The extension was already parsed
+            return Err(X509Error::DuplicateExtensions)
+        }
+
+        if oid == &p2p_ext_oid {
+            let (public_key, signature): (Vec<u8>, Vec<u8>) = yasna::decode_der(ext.value)
+                .map_err(|_| X509Error::InvalidUserCertificate)?;
+            let public_key = identity::PublicKey::from_protobuf_encoding(&public_key)
+                .map_err(|_| X509Error::InvalidUserCertificate)?;
+            let ext = P2pExtension { public_key, signature};
+            libp2p_extension = Some(ext)
```

Comment:
The extension MAY be critical, so we have to `continue` in order to not to hit the next condition `if ext.critical`